File: Chapter 10/Java/src/main/java/com/grokkingsoftwarearchitecture/chapter10/section_10_3_resilience_decorators/resilient/core/application/CheckoutOrchestrator.java
package com.grokking.resilience.core.application;

import com.grokking.resilience.core.domain.OrderStatus;
import com.grokking.resilience.core.ports.MessageQueue;
import com.grokking.resilience.core.ports.PaymentGateway;
import java.util.UUID;

/**
 * THE CORE APPLICATION LAYER:
 * * DESIGN NOTE:
 * This is the "Brain" of the hexagon. It coordinates the business flow 
 * using Ports. It doesn't know about Resilience4j; it only knows 
 * the 'Policy' for success and the 'Plan B' for failure.
 */
public class CheckoutOrchestrator {
    private final PaymentGateway paymentPort;
    private final MessageQueue queuePort;

    public CheckoutOrchestrator(PaymentGateway paymentPort, MessageQueue queuePort) {
        this.paymentPort = paymentPort;
        this.queuePort = queuePort;
    }

    public OrderStatus processCheckout(String orderId, double amount) {
        // #G: IDEMPOTENCY KEY (Generated in Core Application)
        // CRITICAL: Key remains constant across all retry attempts in the adapter.
        String idempotencyKey = UUID.randomUUID().toString();

        try {
            // 1. THE HAPPY PATH (Hidden retries happen inside the adapter)
            paymentPort.charge(amount, orderId, idempotencyKey);
            System.out.println("      [Core Application] Transaction successful.");
            return OrderStatus.PAID;
            
        } catch (Exception e) {
            // #H: THE FALLBACK (Plan B)
            System.out.println("      [Core Application] PRIMARY FAILED. Executing Plan B.");
            
            queuePort.enqueue(orderId, amount, OrderStatus.PENDING_PAYMENT, idempotencyKey);
            return OrderStatus.PENDING_PAYMENT;
        }
    }
}

File: Chapter 10/C#/Section_10_3_Resilience_Decorators/Resilient/Core/Application/CheckoutOrchestrator.cs
using System;
using System.Threading.Tasks;
using Chapter10.Resilience.Core.Domain;
using Chapter10.Resilience.Core.Ports;

namespace Chapter10.Resilience.Core.Application;

public class CheckoutOrchestrator
{
    private readonly IPaymentGateway _paymentGateway;
    private readonly IMessageQueue _messageQueue;

    public CheckoutOrchestrator(IPaymentGateway paymentGateway, IMessageQueue messageQueue)
    {
        _paymentGateway = paymentGateway;
        _messageQueue = messageQueue;
    }

    public async Task<OrderStatus> ProcessCheckoutAsync(string orderId, decimal amount)
    {
        // #G: Idempotency Key is a Business-Level safety mechanism.
        // It must remain constant across all retry attempts.
        string idempotencyKey = Guid.NewGuid().ToString();

        try
        {
            // 1. THE HAPPY PATH (Hidden retries happen inside the Adapter)
            await _paymentGateway.ChargeAsync(amount, orderId, idempotencyKey);
            Console.WriteLine("      [Core Application] PRIMARY SUCCESS: Transaction PAID.");
            return OrderStatus.Paid;
        }
        catch (Exception ex)
        {
            // #H: THE FALLBACK (Plan B)
            // When the adapter's Resilience Policy (Polly) finally gives up, 
            // the Orchestrator executes the recovery path.
            Console.WriteLine($"      [Core Application] PRIMARY FAILED: {ex.Message}");
            Console.WriteLine("      [Core Application] EXECUTING PLAN B: Securing data in Queue.");

            await _messageQueue.EnqueueAsync(new {
                OrderId = orderId,
                Total = amount,
                Status = OrderStatus.PendingPayment,
                Key = idempotencyKey
            });

            return OrderStatus.PendingPayment;
        }
    }
}